File: 5-Problem-A-Day/281. Zigzag Iterator.java
public class ZigzagIterator {

    int idx1, idx2;
    List<Integer> v1, v2;
    boolean one;

    public ZigzagIterator(List<Integer> v1, List<Integer> v2) {
        this.v1 = v1;
        this.v2 = v2;
        idx1 = idx2 = 0;
        one = true;
    }

    public int next() {
        int ans = 0;
        if (idx2 >= v2.size() || (idx1 < v1.size() && one)) {
            ans = v1.get(idx1);
            idx1++;
        } else if (idx1 >= v1.size() || (idx2 < v2.size() && !one)) {
            ans = v2.get(idx2);
            idx2++;
        }
        one = !one;
        return ans;
    }

    public boolean hasNext() {
        return idx1 < v1.size() || idx2 < v2.size();
    }
}

/**
 * Your ZigzagIterator object will be instantiated and called as such: ZigzagIterator i = new
 * ZigzagIterator(v1, v2); while (i.hasNext()) v[f()] = i.next();
 */

File: 2021/281. Zigzag Iterator.cpp
class ZigzagIterator {
public:
    vector<vector<int>> vs;
    int index[2], curr;    
    ZigzagIterator(vector<int>& v1, vector<int>& v2) {
        vs.push_back(v1);
        vs.push_back(v2);
        index[0] = index[1] = 0;
        curr = 0;        
    }

    int next() { 
        if (!hasNext()) {
            return -1;
        }
        int ret = vs[curr][index[curr]];
        index[curr] += 1;
        curr = 1 - curr;
        return ret;
    }

    bool hasNext() {        
        if (index[0] >= vs[0].size() && index[1] >= vs[1].size()) return false;
        if (index[curr] >= vs[curr].size()) {
            curr = 1 - curr;
        }
        return true;
    }
};

/**
 * Your ZigzagIterator object will be instantiated and called as such:
 * ZigzagIterator i(v1, v2);
 * while (i.hasNext()) cout << i.next();
 */